DOC: Add instructions how to activate virtual env under windows with pip

gbaychev opened the pull request:
Contributing.rst now correctly describes how to setup dev environment
under Windows when using pip and not conda.

Closes #29112

- [x] closes #29112 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

gbaychev committed:
DOC: Add instructions how to activate virtual env under windows with pip

Contributing.rst now correctly describes how to setup dev environment
under Windows

Closes #29112

WillAyd commented on doc/source/development/contributing.rst:
```diff
@@ -245,6 +245,10 @@ You'll need to have at least python3.5 installed on your system.
    # Any parent directories should already exist
    python3 -m venv ~/virtualenvs/pandas-dev
    # Activate the virtualenv
+   #    If you are using Windows and Powershell you need to run:
```
Are you sure of these? The virtualenv documentation doesn't make mention of these:

https://virtualenv.pypa.io/en/stable/userguide/#activate-script

Generally a little wary of providing instructions like these for third party libraries as our documentation could easily get out of sync. Maybe just link there instead?

WillAyd replied:
Ah sorry read the documentation in more detail and see it now. I still think preferable just to link there instead though, as it looks like there could be additional nuances

gbaychev replied:
Ah, yes, absolutely correct. I totally forgot about the code execution policies of Powershell.

Is it OK in your opinion, if I write something along the lines: 'If you are using Windows, the activation scripts are located in ~/virtualenvs/pandas-dev/Scripts, for details please refer to: https://virtualenv.pypa.io/en/stable/userguide/#activate-script'

WillAyd commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -195,6 +195,7 @@ Documentation Improvements
 
 - Added new section on :ref:`scale` (:issue:`28315`).
 - Added sub-section Query MultiIndex in IO tools user guide (:issue:`28791`)
+- Added hints how setup a dev environment under Windows when using pip only (:issue:`29112`)
```
This is a nice idea but OK to remove - we don't add whatsnew entries for documentation-only updates

gbaychev replied:
Alright, I was not sure whether or not should add it, will remove it. :)

gbaychev committed:
Remove the unneeded whatsnew entry

gbaychev committed:
Refer to the official user guide regarding the activation of virtual env

gbaychev commented:
Hi @WillAyd ,
I've adjusted the comments as discussed above, please take a look when it is convenient to you

Regards

WillAyd commented:
@datapythonista mind taking a look here?

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -243,8 +243,16 @@ You'll need to have at least python3.5 installed on your system.
+   #     ~\virtualenvs\pandas-dev\scripts
+   # Please refer to the official user guide at
+   # https://virtualenv.pypa.io/en/stable/userguide/#activate-script
+   # about how to activate your virtual environment under Windows
```
Thanks for clarifying this @gbaychev.

Don't you think could be clearer to have the windows instructions in a separate block? Feels like there is too much information in this block now. And having two blocks would let users just read the one they care about, and have the information presented in a clearer way. 

Feel free to disagree, but I think we can probably do better at communicating this.

gbaychev replied:
You guys are very careful and diplomatic, cheers for that :) Yes @datapythonista,  you are absolutely correct, I've split the information into two code-blocks, please see 39f6fa7

gbaychev committed:
Move the Windows doc about venv with pip into a separate code block

Relevant issue #29112

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -254,6 +257,29 @@ You'll need to have at least python3.5 installed on your system.
+**Windows**
+
+Below is a brief overview on how to set-up a virtual environment with Powershell
+under Windows. For details please refer to the \
```
Do we need the backslash?

gbaychev replied:
Actually, no, removed it.

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -254,6 +257,29 @@ You'll need to have at least python3.5 installed on your system.
+under Windows. For details please refer to the \
+`official virtualenv user guide <https://virtualenv.pypa.io/en/stable/userguide/#activate-script>`__
+
+.. code-block:: none
```
I think using `bash` instead of `none` should highlight this in a nice way (comments in a different color). I know you copied it from the previous block, but I think you can change that too.

gbaychev replied:
Added bash to the previous block, powershell to the Windows one.

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -254,6 +257,29 @@ You'll need to have at least python3.5 installed on your system.
+   python -m pip install -r requirements-dev.txt
+
+   # Build and install pandas
+   python setup.py build_ext --inplace -j 4
```
not sure why we had `-j 4` initially, but I think it's better `-j auto`, so all cores are used.

gbaychev replied:
At least under Windows you need to use an integer for -j, 4 is a good value IMHO. 

![Capture](https://user-images.githubusercontent.com/3358056/67565120-f1da2700-f724-11e9-9ce4-0ff3b7fec2b5.PNG)

datapythonista reviewed (commented):
I'm a bit confused. Your previous diff was quite big, but if I'm not missing something, now the only difference is the activation. Is there something missing? If it's just that, just a single block makes more sense, and that line could simply be changing this in this PR:
```
. ~/virtualenvs/pandas-dev/bin/activate (~\virtualenvs\pandas-dev\Scripts\Activate.ps1 in Windows)
```

Commit:
Additional refinement

- add aditional information about cmd.exe
- use $env:Userprofile where appropriate
- add --use-no-pep517 to the windows instruction
- add some syntax highlight

Relevant issue #29113

gbaychev commented:
Alright I need to apologize - yesterday I was tired and I've committed some brain fart to the repo.
What was missing was the information about the cmd.exe. Under Windows you can use two shells (WSL is a whole other story) - cmd.exe and powershell and they behave quite differently.

Now I've added the info about cmd.exe in the comments in the Windows code-block. Also I did some more testing with powershell and I've adjusted the initial generation of the virtual env, so that it 100% functions.

I've also added the `--no-use-pep517` argument to the Windows instructions, because it seems that under Windows a pep517 build gets triggered always. At least on the three different machines, I've tested with. Please see #28633

datapythonista reviewed (commented):
Couple of small things, and you need to fix the CI (tailing whitespaces).

Should be good with that.

gbaychev committed:
PR improvements

- remove trailing spacing remove
- change the setup.py params so that all CPU cores will be utilized
- split cmd.exe instructions on a separate line

Related issue #29112

gbaychev commented on doc/source/development/contributing.rst:
```diff
@@ -254,6 +257,34 @@ You'll need to have at least python3.5 installed on your system.
+
+   # Activate the virtualenv
+   # If you are using cmd.exe, run instead: %USERPROFILE%\virtualenvs\pandas-dev\Scripts\activate.bat
+   ~\virtualenvs\pandas-dev\Scripts\Activate.ps1 
```
Done

gbaychev commented:
Hello, I did the requested changes

Regards

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -238,22 +238,55 @@ Creating a Python environment (pip)
+   # Use an ENV_DIR of your choice. We'll use ~\virtualenvs\pandas-dev where
+   # '~' is the folder pointed to by either $env:USERPROFILE (Powershell) or
+   # %USERPROFILE% (cmd.exe) environment variable
+   # Any parent directories should already exist
```
I'd move this outside of the code block, it'll be easier to read as a paragraph than as code comments.

Also, since you explain the equivalence of `$env:USERPROFILE` and `%USERPROFILE%` here, you can just leave one of them below. Will look much cleaner and clearer without duplicating the ocmmands.

Other than that, lgtm

gbaychev committed:
Extract some windows pip venv comment insturctions into a separate paragraph

datapythonista commented on doc/source/development/contributing.rst:
```diff
@@ -238,22 +238,51 @@ Creating a Python environment (pip)
+   python -m venv $env:USERPROFILE\virtualenvs\pandas-dev
+
+   # Activate the virtualenv. Use activate.bat for cmd.exe
+   ~\virtualenvs\pandas-dev\Scripts\Activate.ps1
```
I see that we use `$env:USERPROFILE` in the previous command, and `~` here. Is this correct?

gbaychev replied:
Yes, this is correct. Powershell on its own resolves the ' ~ ' to correctly to the home directory. But I noticed, that  if I call `python -m venv ~\virtualenvs\pandas-dev`, venv won't expand ' ~ ' to your home folder and will create a folder called '~' in your current working directory. Hence I've added $env:USERPROFILE, it is the only place where is really needed, if you are using powershell. 

Since ' ~ ' is explained in the now extracted paragraph above, I think the snippet you've quoted can stay as it is.

datapythonista reviewed (approved):
lgtm, @WillAyd do you want to have a look?

WillAyd commented:
Thanks @gbaychev
